cmd/catalyst-node/catalyst-node.go

```diff
@@ -751,3 +760,203 @@ func queryMistForClosestNodeSource(playbackID, lat, lon, prefix string, source b
+
+		if time.Now().After(expiration) {
+			glog.Errorf("Token expired %v", expiration)
+			return []byte("false")
+		}
+	}
+
+	playbackAccessControl, err := ac.getPlaybackAccessControlInfo(playbackID, pubKey)
+	if err != nil {
+		glog.Errorf("Unable to get playback access control info %v", err)
+		return []byte("false")
+	}
+
+	if playbackAccessControl[pubKey].allow {
+		return []byte("true")
+	}
+
+	return []byte("false")
+}
+
+func (ac *PlaybackAccessControl) getPlaybackAccessControlInfo(playbackID, pubKey string) (map[string]*PlaybackAccessControlEntry, error) {
+	if ac.cache[playbackID] == nil ||
+		ac.cache[playbackID][pubKey] == nil ||
+		time.Now().After(ac.cache[playbackID][pubKey].Stale) {
+
+		err := ac.cachePlaybackAccessControlInfo(playbackID, pubKey)
+		if err != nil {
+			return nil, err
+		}
+
+	} else if time.Now().After(ac.cache[playbackID][pubKey].MaxAge) {
+		go ac.cachePlaybackAccessControlInfo(playbackID, pubKey)
+	}
+
+	return ac.cache[playbackID], nil
+}
+
+func (ac *PlaybackAccessControl) cachePlaybackAccessControlInfo(playbackID, pubKey string) error {
+	body, err := json.Marshal(PlaybackAccessControlRequest{"jwt", pubKey, playbackID})
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", ac.gateURL, bytes.NewReader(body))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err := ac.Client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer res.Body.Close()
+
+	cacheControlDirectives := strings.Split(res.Header.Get("Cache-Control"), ",")
+
+	var maxAge, stale int64 = 0, 0
+	if len(cacheControlDirectives) == 2 {
+		maxAge, err = strconv.ParseInt(strings.Split(cacheControlDirectives[0], "=")[1], 10, 32)
+		if err != nil {
+			return err
+		}
+
+		stale, err = strconv.ParseInt(strings.Split(cacheControlDirectives[1], "=")[1], 10, 32)
+		if err != nil {
+			return err
+		}
+	}
+
+	var allow = res.StatusCode == 204
+	var maxAgeTime = time.Now().Add(time.Duration(maxAge) * time.Second)
+	var staleTime = time.Now().Add(time.Duration(stale) * time.Second)
+
+	if ac.cache[playbackID] == nil {
+		ac.cache[playbackID] = make(map[string]*PlaybackAccessControlEntry)
+	}
+
+	ac.cache[playbackID][pubKey] = &PlaybackAccessControlEntry{staleTime, maxAgeTime, allow}
+
+	return nil
+}
+
+type GatedStreamClaims struct {
+	Action string `json:"action"`
+	Video  string `json:"video"`
+	Pub    string `json:"pub"`
+	jwt.RegisteredClaims
+}
+
+func decodeJwt(tokenString string) (jwt.Claims, error) {
+	token, _ := jwt.Parse(tokenString, nil)
+
+	pub := token.Claims.(jwt.MapClaims)["pub"].(string)
+	decodedPubkey, err := base64.StdEncoding.DecodeString(pub)
+	if err != nil {
+		return nil, err
+	}
+
+	publicKey, err := jwt.ParseECPublicKeyFromPEM(decodedPubkey)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+		return publicKey, nil
+	})
```

Comment:
I believe we don't need to parse the token twice. The function provided to `Parse` is made for fetching the public key based on the parsed token, so we should do the `pub` parsing inside there.

I also think that this logic would be greatly improved if we used the `ParseWithClaims` version: https://pkg.go.dev/github.com/golang-jwt/jwt/v4#example-ParseWithClaims-CustomClaimsType
Which avoids the use of `interface{}` casting around the code which can cause panics.

I played around with it a bit on top of that example for signing a JWT, and came out with this: https://gist.github.com/victorges/d9645525caa89945012dafe8ac3bb3fc#file-jwt-validation-go-L77

WDYT? This also keeps a lot of the lower level validation inside the claims `Valid()` function, which also makes sense to me as it separates concerns of protocol vs contextual validations.